lightning/src/onion_message/messenger.rs

```diff
@@ -290,6 +290,16 @@ impl<Signer: Sign, K: Deref, L: Deref> OnionMessageProvider for OnionMessenger<S
 	      L::Target: Logger,
 {
 	fn next_onion_messages_for_peer(&self, peer_node_id: PublicKey, max_messages: usize) -> Vec<msgs::OnionMessage> {
+		let mut pending_msgs = self.pending_messages.lock().unwrap();
+		if let Some(msgs) = pending_msgs.get_mut(&peer_node_id) {
+			if max_messages >= msgs.len() {
+				let mut peer_pending_msgs = Vec::new();
+				mem::swap(msgs, &mut peer_pending_msgs);
+				return peer_pending_msgs
+			} else {
+				return msgs.split_off(max_messages)
+			}
+		}
 		Vec::new()
 	}
```

Comment:
Not sure how much concern there is around heap allocation, but if so we could have the interface return `Option<msgs::OnionMessage>` and implement it using a `VecDeque`. It would simplify this code to:

```rust
self.pending_messages.lock().unwrap().get_mut(&peer_node_id).and_then(|msgs| msgs.pop_front())
```

Reply:
I talked to Matt and apparently because the `OnionMessenger` already maintains a vec of onion messages, the current implementation won't actually allocate very often, so it isn't worth it to return one `Optional` OM at a time. 

The idea is that the `split_off` case should be rare, and the other cases don't allocate because `Vec::new()` doesn't allocate.

Reply:
Won't it allocate essentially every time you receive an onion message since any call to `next_onion_messages_for_peer` will swap the current `Vec` with the empty `Vec`, which upon push will need to allocate memory since it has a capacity of zero?

Reply:
For context here -- in #1660 the reasoning for buffering gossip one at a time was "... only buffer one message at a time, as the peers' outbound socket buffer drains. This moves the outbound backfill messages out of PeerHandler and into the operating system buffer, where it arguably belongs." 

@TheBlueMatt I'm wondering why this reasoning doesn't apply to onion messages? Is it less reliable, and we're looser about dropping gossip?

Reply:
> Its kinda awkward to run it in a tight loop, though, as we'll end up locking and unlocking the onion messenger peer buffer a ton, which seems annoying? Can go either way here, as far as I'm concerned.

FWIW, it looks like we are doing that anyway with a different mutex when calling `get_peer_for_forwarding!` when looping over the messages.

Reply:
Not sure I follow. What I was saying is even if we didn't change `next_onion_messages_for_peer`, we then immediately iterate over the onion messages in the returned `Vec` and lock/unlock the peer mutex on each iteration here:

https://github.com/lightningdevkit/rust-lightning/blob/652d54a67e3b876584441c7a636d78a210888d03/lightning/src/ln/peer_handler.rs#L1705

Reply:
I moved it into the write method. Let me know if this looks close to good, would like confirmation that the consts we had prior are no longer relevant before resolving some review on them